File: Data Structures & Algorithms/longest-increasing-subsequence/submission-0.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        
        int len = nums.size();
        vector<int> dp(len+1, 1);

        int maxLength = 0;
        for(int i=0;i<len;i++)
        {
            for(int j=0;j<i;j++)
            {
                if(nums[j] < nums[i]) dp[i] = max(dp[i], 1 + dp[j]);
            }

            maxLength = max(maxLength, dp[i]);
        }

        return maxLength;
    }
};

File: Data Structures & Algorithms/top-k-elements-in-list/submission-0.js
class Solution {
    /**
     * @param {number[]} nums
     * @param {number} k
     * @return {number[]}
     */
    topKFrequent(nums, k) {
        const freqMap = {};

        for(const num of nums)
        {
            if(!freqMap[num]) freqMap[num] = 1;
            else freqMap[num] += 1;
        }

        let bucket = [];
        for(let key in freqMap)
        {
            const freq = freqMap[key];
            if(!bucket[freq]) bucket[freq] = [];
            bucket[freq].push(Number(key));
        }

        const result = [];
        for(let i=bucket.length - 1; i >= 0; i--)
        {
            if(bucket[i]) result.push(...bucket[i]);
            if(result.length == k) break;
        }

        return result;
    }
}
